### PR TITLE
feat: new controller poc

### PR DESCRIPTION
**Pull request by yannongli-okta**
## Description:
POC for new controller proposal

## PR Checklist

- [ ] Have you verified the basic functionality for this change?

- [ ] Added unit tests?

- [ ] Added e2e tests

- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

### Issue:

- [OKTA-XXXXXX](https://oktainc.atlassian.net/browse/OKTA-XXXXXX)

### CODE REVIEW

**Review by pradeepdewda-okta — COMMENTED** [review]
Thanks for making changes as per my recommendation. We can improve it from here.

File: src/v2/controllers/DeviceEnrollmentTerminalViewController.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import { _, Controller } from 'okta';
+import ViewFactory from '../view-builder/ViewFactory';
+import {FORMS as RemediationForms} from '../ion/RemediationConstants';
+
+export default Controller.extend({
+  className: 'deviceEnrollment-controller',
+
```
Comment (pradeepdewda-okta) [inline src/v2/controllers/DeviceEnrollmentTerminalViewController.js:7]: set a flag that say this controller does not have contain any idx remediation responses and it is a terminal state.

`terminalStateWithNoRemediation : true`

File: src/v2/BaseLoginRouter.js [hunk 2]
```diff
@@ -151,17 +171,21 @@ export default Router.extend({
...
-          this.settings.unset('stateToken');
-          this.appState.trigger('remediationSuccess', idxResp);
-          this.render(Controller, options);
-        })
-        .catch(errorResp => {
-          this.settings.unset('stateToken');
-          this.appState.trigger('remediationError', errorResp.error);
-          this.render(Controller, options);
-        });
+      if (this.settings.options.deviceEnrollment && this.settings.options.deviceEnrollment.name) {
```
Comment (pradeepdewda-okta) [inline src/v2/BaseLoginRouter.js:174]: remove this `if`.. and all terminal state with no remediation should be part of else statement.

File: src/v2/WidgetRouter.js [hunk 3]
```diff
@@ -12,15 +12,21 @@
...
     '*wildcard': 'defaultAuth',
+    'authenticators/ov-not-installed': 'deviceEnrollmentTerminalView'
   },
 
   defaultAuth: function () {
     this.render(FormController);
   },
 
+  deviceEnrollmentTerminalView: function () {
+    this.render(DeviceEnrollmentTerminalViewController);
```
Comment (pradeepdewda-okta) [inline src/v2/WidgetRouter.js:29]: pass options in render that set a flag. Meaning this controller does not contain any idx remediation responses and it is a terminal state.

`this.render(DeviceEnrollmentTerminalViewController, { terminalStateWithNoRemediation : true })`

File: src/v2/BaseLoginRouter.js (hunk 2 above)
Comment (pradeepdewda-okta) [reply src/v2/BaseLoginRouter.js:174]: use `terminalStateWithNoRemediation` that passed as options to check for no remediation terminal state..

basically
```
if (this.settings.get('stateToken') && !options.terminalStateWithNoRemediation`) 
--- idx flow
else 
--- terminal states
```

File: src/v2/BaseLoginRouter.js [hunk 4]
```diff
@@ -130,6 +130,26 @@ export default Router.extend({
     }
   },
 
+  handleDeviceEnrollmentFlow () {
+    const deviceEnrollmentValue = this.settings.options.deviceEnrollment;
```
Comment (pradeepdewda-okta) [inline src/v2/BaseLoginRouter.js:134]: pass these as options from widgetRouter.. same as the new flag.

File: src/v2/BaseLoginRouter.js [hunk 5]
```diff
@@ -130,6 +130,26 @@ export default Router.extend({
...
+      version: '1.0.0',
+      expiresAt: '2020-06-14T22:15:50.000Z',
+      intent: 'LOGIN',
+      deviceEnrollment: {
+        type: 'object',
+        value: deviceEnrollmentValue
+      }
+    };
+
+    this.handleIdxResponseSuccess({
```
Comment (pradeepdewda-okta) [inline src/v2/BaseLoginRouter.js:146]: create a new event  `terminalStateWithNoRemediation ` and listen to it at line 82
and handler `handleTerminalStateWithNoRemediation`

File: src/v2/BaseLoginRouter.js [hunk 6]
```diff
@@ -151,17 +171,21 @@ export default Router.extend({
...
-          this.appState.trigger('remediationSuccess', idxResp);
-          this.render(Controller, options);
-        })
-        .catch(errorResp => {
-          this.settings.unset('stateToken');
-          this.appState.trigger('remediationError', errorResp.error);
-          this.render(Controller, options);
-        });
+      if (this.settings.options.deviceEnrollment && this.settings.options.deviceEnrollment.name) {
+        this.handleDeviceEnrollmentFlow();
```
Comment (pradeepdewda-okta) [inline src/v2/BaseLoginRouter.js:175]: trigger `terminalStateWithNoRemediation` event

```
this.appState.trigger('terminalStateWithNoRemediation', options);
```

File: src/v2/BaseLoginRouter.js [hunk 7]
```diff
@@ -130,6 +142,50 @@ export default Router.extend({
     }
   },
 
+  handleTerminalStateWithNoRemediation () {
+    const deviceEnrollmentType = this.settings.get('deviceEnrollment.name');
+    let deviceEnrollmentResponse;
+
+    if (deviceEnrollmentType === 'oda') {
+      deviceEnrollmentResponse = {
```
Comment (pradeepdewda-okta) [inline src/v2/BaseLoginRouter.js:150]: remove it from here and pass as options in ``deviceEnrollmentTerminalView` function in WidgetRouter.js

File: src/v2/BaseLoginRouter.js [hunk 8]
```diff
@@ -130,6 +142,50 @@ export default Router.extend({
...
+          type: 'object',
+          value: {
+            name: deviceEnrollmentType,
+            platform: this.settings.get('deviceEnrollment.platform'),
+            signInUrl: this.settings.get('deviceEnrollment.signInUrl')
+          }
+        }
+      };
+    } else {
+      deviceEnrollmentResponse = {
```
Comment (pradeepdewda-okta) [inline src/v2/BaseLoginRouter.js:165]: same here as well.

File: test/unit/spec/DeviceEnrollmentTerminalView_spec.js [hunk 9]
```diff
@@ -0,0 +1,89 @@
+import { _ } from 'okta';
```
Comment (pradeepdewda-okta) [inline test/unit/spec/DeviceEnrollmentTerminalView_spec.js:1]: the unit test goes to test/unit/spec/v2

File: src/v2/WidgetRouter.js [hunk 10]
```diff
@@ -12,15 +12,37 @@
...
   },
 
   defaultAuth: function () {
     this.render(FormController);
   },
 
+  deviceEnrollmentTerminalView: function () {
+    this.render(DeviceEnrollmentTerminalViewController, {
+      terminalStateWithNoRemediation : true,
+      deviceEnrollmentTerminalResponse: {
```
Comment (pradeepdewda-okta) [inline src/v2/WidgetRouter.js:31]: its a big object, create a variable and then use it here.

```
let response = { ... }
```

File: src/models/Settings.js [hunk 11]
```diff
@@ -105,6 +105,13 @@ export default Model.extend({
       value: 'SECONDARY',
     },
 
+    // Device Enrollment Remediation
```
Comment (pradeepdewda-okta) [inline src/models/Settings.js:108]: I doubt we need to change Settings.js from V1 code base.. All of your changes should go to V2 only..

Device enrollment object should be available in appState.. @jhe-okta has added it already..
https://github.com/okta/okta-signin-widget/blob/master/src/v2/ion/responseTransformer.js#L177

Comment (yannongli-okta) [reply src/models/Settings.js:108]: Right. I just tested locally and it works without adding these in Setting.js. Will remove in next commit.

**Comment by yannongli-okta** [comment]
Official PR has been out

**Pull request closed without merging**